lib/services/role.go

```diff
@@ -541,43 +971,78 @@ func (set RoleSet) CheckLogins(ttl time.Duration) ([]string, error) {
+// checked first then allow rules. Access to a node is determined by
+// namespaces, labels, and logins.
 func (set RoleSet) CheckAccessToServer(login string, s Server) error {
 	var errs []error
 
+	// check deny rules first: a single matching namespace, label, or login from
+	// the deny role set prohibits access.
+	for _, role := range set {
+		matchNamespace := MatchNamespace(role.GetNamespaces(Deny), s.GetNamespace())
+		matchLabels := MatchLabels(role.GetNodeLabels(Deny), s.GetAllLabels())
+		matchLogin := MatchLogin(role.GetLogins(Deny), login)
+		if matchNamespace || matchLabels || matchLogin {
+			errorMessage := fmt.Sprintf("Role %v denied access to node %v: deny rule matched; match(namespace=%v, label=%v, login=%v)",
+				role.GetName(), s.GetHostname(), matchNamespace, matchLabels, matchLogin)
+			log.Warnf("[RBAC] " + errorMessage)
+			return trace.AccessDenied(errorMessage)
+		}
+	}
+
+	// check allow rules: namespace, label, and login have to all match in
+	// one role in the role set to be granted access.
 	for _, role := range set {
-		matchNamespace := MatchNamespace(role.GetNamespaces(), s.GetNamespace())
-		matchLabels := MatchLabels(role.GetNodeLabels(), s.GetAllLabels())
-		matchLogin := MatchLogin(role.GetLogins(), login)
+		matchNamespace := MatchNamespace(role.GetNamespaces(Allow), s.GetNamespace())
+		matchLabels := MatchLabels(role.GetNodeLabels(Allow), s.GetAllLabels())
+		matchLogin := MatchLogin(role.GetLogins(Allow), login)
 		if matchNamespace && matchLabels && matchLogin {
-			log.Debugf("Role %v granted access to node %v", role.GetName(), s.GetHostname())
+			log.Debugf("[RBAC] Role %v granted access to node %v", role.GetName(), s.GetHostname())
 			return nil
 		}
 
-		errorMessage := fmt.Sprintf("Role %v denied access; match(namespace=%v, label=%v, login=%v)",
+		errorMessage := fmt.Sprintf("Role %v denied access: allow rules did not match; match(namespace=%v, label=%v, login=%v)",
 			role.GetName(), matchNamespace, matchLabels, matchLogin)
 		errs = append(errs, trace.AccessDenied(errorMessage))
 	}
 
 	return trace.AccessDenied("access to node %v is denied to role(s): %v", s.GetHostname(), errs)
 }
 
-// CanForwardAgents returns true if role set allows forwarding agents
+// CanForwardAgents returns true if role set allows forwarding agents.
 func (set RoleSet) CanForwardAgents() bool {
 	for _, role := range set {
-		if role.CanForwardAgent() {
+		forwardAgent, err := role.GetOptions().GetBoolean(ForwardAgent)
+		if err != nil {
+			return false
+		}
+		if forwardAgent == true {
 			return true
 		}
 	}
 	return false
 }
 
-// CheckAgentForward checks if the role can request agent forward for this user
+// CheckAgentForward checks if the role can request to forward the SSH agent
+// for this user.
 func (set RoleSet) CheckAgentForward(login string) error {
+	// deny check: check if role we have permission to login, can't forward connections
```

Comment:
this seems excessive, it mostly duplicates the other check in the same role `CheckLoginDuration` AFAIK, I would try to avoid logic duplication